File: backend/src/phasor_workbench/validation/paths.py
"""Locating a problem within the submitted document.

Paths use the format documented in docs/API.md -- `operations[4].targets[0]` --
and are built from **wire** field names, not Python attribute names. The
generated models rename `classicalTarget` to `classical_target` and alias it
back on the wire; a path naming the Python attribute would point at a field the
client never sent.
"""

from __future__ import annotations

from collections.abc import Iterator

from ..models import Operation
from ..models.circuit import GateOperation


def operation(index: int) -> str:
    return f"operations[{index}]"


def qubit(index: int, field: str) -> str:
    return f"qubits[{index}].{field}"


def register(index: int, field: str) -> str:
    return f"classicalRegisters[{index}].{field}"


def qubit_references(op: Operation, index: int) -> Iterator[tuple[str, str]]:
    """Yield `(qubit id, path)` for every qubit an operation names.

    Targets first, then controls, in document order. Only gates have controls;
    the schema forbids them on measurements and barriers, so a parsed operation
    of either kind has none to yield.
    """
    for position, target in enumerate(op.targets):
        yield target, f"{operation(index)}.targets[{position}]"

    if isinstance(op, GateOperation):
        for position, control in enumerate(op.controls):
            yield control, f"{operation(index)}.controls[{position}]"

File: frontend/src/validation/paths.ts
/**
 * Locating a problem within the submitted document.
 *
 * Mirrors `backend/src/phasor_workbench/validation/paths.py`, and must produce
 * byte-identical paths for the same defect. Format is documented in
 * docs/API.md: `operations[4].targets[0]`.
 *
 * The frontend has an easier time of this than Python, whose generated models
 * rename `classicalTarget` and alias it back on the wire. Here the field names
 * already are the wire names.
 */

import type { Operation } from '../model/circuit';

export function operation(index: number): string {
  return `operations[${String(index)}]`;
}

export function qubit(index: number, field: string): string {
  return `qubits[${String(index)}].${field}`;
}

export function register(index: number, field: string): string {
  return `classicalRegisters[${String(index)}].${field}`;
}

export interface QubitReference {
  readonly id: string;
  readonly path: string;
}

/**
 * Every qubit an operation names: targets first, then controls, in document
 * order.
 *
 * `controls` is optional in the generated TypeScript but defaulted in Python,
 * so this is one of the few places the two bindings genuinely differ in shape.
 * The empty fallback is what keeps the two implementations agreeing.
 */
export function qubitReferences(
  op: Operation,
  index: number,
): QubitReference[] {
  const references: QubitReference[] = op.targets.map((target, position) => ({
    id: target,
    path: `${operation(index)}.targets[${String(position)}]`,
  }));

  if (op.kind === 'gate') {
    const controls: readonly string[] = op.controls ?? [];
    controls.forEach((control, position) => {
      references.push({
        id: control,
        path: `${operation(index)}.controls[${String(position)}]`,
      });
    });
  }

  return references;
}
